packages/expo-file-system/android/src/main/java/expo/modules/filesystem/FileSystemModule.kt

```diff
@@ -0,0 +1,1377 @@
+        } else {
+          if (options.containsKey("idempotent") && options["idempotent"] as Boolean) {
+            promise.resolve(null)
+          } else {
+            promise.reject(
+              "ERR_FILESYSTEM_CANNOT_FIND_FILE",
+              "File '$uri' could not be deleted because it could not be found"
+            )
+          }
+        }
+      } else {
+        throw IOException("Unsupported scheme for location '$uri'.")
+      }
+    } catch (e: Exception) {
+      e.message?.let { Log.e(TAG, it) }
+      promise.reject(e)
+    }
+  }
+
+  @ExpoMethod
+  fun moveAsync(options: Map<String?, Any?>, promise: Promise) {
+    try {
+      if (!options.containsKey("from")) {
+        promise.reject("ERR_FILESYSTEM_MISSING_PARAMETER", "`FileSystem.moveAsync` needs a `from` path.")
+        return
+      }
+      val fromUri = Uri.parse(options["from"] as String?)
+      ensurePermission(Uri.withAppendedPath(fromUri, ".."), Permission.WRITE, "Location '$fromUri' isn't movable.")
+      if (!options.containsKey("to")) {
+        promise.reject("ERR_FILESYSTEM_MISSING_PARAMETER", "`FileSystem.moveAsync` needs a `to` path.")
+        return
+      }
+      val toUri = Uri.parse(options["to"] as String?)
+      ensurePermission(toUri, Permission.WRITE)
+      if (fromUri.scheme == "file") {
+        val from = uriToFile(fromUri)
+        val to = uriToFile(toUri)
+        if (from.renameTo(to)) {
+          promise.resolve(null)
+        } else {
+          promise.reject(
+            "ERR_FILESYSTEM_CANNOT_MOVE_FILE",
+            "File '$fromUri' could not be moved to '$toUri'"
+          )
+        }
+      } else if (isSAFUri(fromUri)) {
+        val documentFile = getNearestSAFFile(fromUri)
+        if (documentFile == null || !documentFile.exists()) {
+          promise.reject("ERR_FILESYSTEM_CANNOT_MOVE_FILE", "File '$fromUri' could not be moved to '$toUri'")
+          return
+        }
+        val output = File(toUri.path)
+        transformFilesFromSAF(documentFile, output, false)
+        promise.resolve(null)
+      } else {
+        throw IOException("Unsupported scheme for location '$fromUri'.")
+      }
+    } catch (e: Exception) {
+      e.message?.let { Log.e(TAG, it) }
+      promise.reject(e)
+    }
+  }
+
+  @ExpoMethod
+  fun copyAsync(options: Map<String?, Any?>, promise: Promise) {
+    try {
+      if (!options.containsKey("from")) {
+        promise.reject("ERR_FILESYSTEM_MISSING_PARAMETER", "`FileSystem.moveAsync` needs a `from` path.")
+        return
+      }
+      val fromUri = Uri.parse(options["from"] as String?)
+      ensurePermission(fromUri, Permission.READ)
+      if (!options.containsKey("to")) {
+        promise.reject("ERR_FILESYSTEM_MISSING_PARAMETER", "`FileSystem.moveAsync` needs a `to` path.")
+        return
+      }
+      val toUri = Uri.parse(options["to"] as String?)
+      ensurePermission(toUri, Permission.WRITE)
+      when {
+        fromUri.scheme == "file" -> {
```

Comment:
I like it! 🥇 Could you apply this in other places as well? I've seen this in at least two other methods